File: Cpp/Range Sum of BST.cpp
// 144ms, 86.45%; 41.1MB, 92.73%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int rangeSumBST(TreeNode *root, int L, int R)
    {
        stack<TreeNode *> st;
        st.push(root);
        int _sum = 0;
        while (!st.empty()) { // a bit faster than recursion, but close
            auto tmp = st.top();
            st.pop();
            if (tmp->val >= L && tmp->val <= R) _sum += tmp->val;
            if (tmp->left && tmp->val > L) st.push(tmp->left);
            if (tmp->right && tmp->val < R) st.push(tmp->right);
        }
        return _sum;
    }
};

File: Python3/Range Sum of BST.py
# 228ms, 64.67%; 20.9MB, 100.00%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
        self.sum = 0
        self.__helper(root, L, R)
        return self.sum

    def __helper(self, node: TreeNode, L: int, R: int) -> int:
        if not node or L > R: return
        if L <= node.val <= R: self.sum += node.val
        if node.val > L: self.__helper(node.left, L, min(R, node.val - 1))
        if node.val < R: self.__helper(node.right, max(L, node.val + 1), R)
